src/main/java/org/manifold/compiler/front/Main.java

```diff
@@ -129,172 +118,33 @@ public Schematic invokeFrontend(CommandLine cmd) throws Exception {
       }
     }
 
-    TypeChecker.typecheck(namespaces, defaultNamespace);
+    log.debug("top-level identifiers:");
+    for (VariableIdentifier id :
+        defaultNamespace.getPrivateScope().getSymbolIdentifiers()) {
+      log.debug(id);
+    }
 
-    ExpressionGraph exprGraph = new ExpressionGraph(
-        defaultNamespace.getPrivateScope());
-    exprGraph.buildFrom(expressions);
-    exprGraph.removeUnconnectedEdges();
-    exprGraph.optimizeOutVariables();
+    // let's see if we can do this before static type-checking
+    ExpressionGraphBuilder exprGraphBuilder = new ExpressionGraphBuilder(
+        expressions, namespaces);
+    ExpressionGraph exprGraph = exprGraphBuilder.build();
+    
+    // TODO expression graph correctness checks:
+    // * all variables are assigned exactly once
 
+    log.debug("writing out initial expression graph");
     File exprGraphDot = new File(inputFile.getName() + ".exprs.dot");
     exprGraph.writeDOTFile(exprGraphDot);
-
-    exprGraph.elaboratePrimitives();
-    exprGraph.elaborateConnections(schematic);
-    exprGraph.writeSchematic(schematic);
+    
+    // TODO write out the schematic
```

Comment:
Instead of writing out the schematic, we return it and allow the engine to either pass it to a backend or write it somewhere. However, the correctness checks thing is valid. Opened #51.